### PR TITLE
refactor: remove unneeded ErrorKinds

### PR DESCRIPTION
**Pull request by bartlomieju**
Follow up to #3662

- [x] replace `DenoError.Other` with `Error`
- [x] replace `DenoError.UrlParse`, `DenoError.InvalidPath` and `DenoError.ImportMissingPrefix` with `URIError`
- [ ] replace `DenoError.Http` with `HttpError` (?) - the same error that `fetch` throws
- [ ] replace following with `IoError` (?)
   - `NotFound`
   - `PermissionDenied`
   - `ConnectionRefused`
   - `ConnectionReset`
   - `ConnectionAborted`
   - `NotConnected`
   - `AddrInUse`
   - `AddrNotAvailable`
   - `BrokenPipe`
   - `AlreadyExists`
   - <s>`WouldBlock` - shouldn't be used</s> removed
   - `InvalidInput` - `TypeError`
   - `InvalidData` - <s>`TypeError`</s> (possible just `Error`?)
   - `TimedOut`
   - `Interrupted` - shouldn't be used
   - `WriteZero`
   - `UnexpectedEof`
   - `BadResource`
- [x] replace `InvalidSeekMode` with `TypeError`
- [x] replace `TooLarge` with `Error`
- [x] replace `UnixError` with `SystemError` (?) -> explode that to ERRNO, map ERRNO to Deno defined errors, if one does not exist then panic
- [x] remove `Diagnostic` and `JSError`

### CODE REVIEW

File: cli/deno_error.rs [hunk 1]
```diff
@@ -1,6 +1,18 @@
...
+/// Currently it is deeply intertwined with `ErrBox` which is
+/// not optimal since not every ErrBox can be "JS runtime error";
+/// eg. there's no way to throw JSError/Diagnostic from within JS runtime
+///
+/// TODO:
+///  - rename ErrorKind::Other to WindowError/GenericError
+///  - remove ErrorKind::WouldBlock
+///  - possibly rename "GetErrorKind" to "RuntimeError" - it will allow
+///    better semantic separation of errors which can be sent to JS runtime
+///    eg. each RuntimeError is ErrBox, but not every ErrBox is RuntimeError
```
Comment (ry) [inline cli/deno_error.rs]: Use `//!`

File: cli/deno_error.rs [hunk 2]
```diff
@@ -1,6 +1,18 @@
...
 use crate::import_map::ImportMapError;
 pub use crate::msg::ErrorKind;
```
Comment (ry) [inline cli/deno_error.rs]: I think ErrorKind should be moved this module. It's very related to DenoError.

Comment (bartlomieju) [reply cli/deno_error.rs]: Yeah, `msg.rs` is cruft

File: cli/deno_error.rs [hunk 3]
```diff
@@ -1,6 +1,18 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
-use crate::diagnostics::Diagnostic;
-use crate::fmt_errors::JSError;
+
+/// This module implements error serialization; it
+/// allows to serialize Rust errors to be sent to JS runtime.
+///
+/// Currently it is deeply intertwined with `ErrBox` which is
+/// not optimal since not every ErrBox can be "JS runtime error";
+/// eg. there's no way to throw JSError/Diagnostic from within JS runtime
```
Comment (ry) [inline cli/deno_error.rs]: There are many types of errors in Deno:
- ErrBox: a generic boxed object. This is the super type of all errors handled in Rust.
- JSError: exceptions thrown from V8 into Rust. Usually a user exception. These are basically a big JSON structure which holds information about line numbers. We use this to pretty-print stack traces. These are never passed back into the runtime.
- DenoError: these are errors that happen during ops, which are passed back into the runtime, where an exception object is created and thrown. DenoErrors have an integer code associated with them - access this via the kind() method.
- Diagnostic: these are errors that originate in TypeScript's compiler. They're similar to JSError, in that they have line numbers. But Diagnostics are compile-time type errors, whereas JSErrors are runtime exceptions.

File: cli/deno_error.rs [hunk 4]
```diff
@@ -123,12 +123,7 @@ impl GetErrorKind for ImportMapError {
 
 impl GetErrorKind for ModuleResolutionError {
   fn kind(&self) -> ErrorKind {
-    use ModuleResolutionError::*;
-    match self {
-      InvalidUrl(ref err) | InvalidBaseUrl(ref err) => err.kind(),
-      InvalidPath(_) => ErrorKind::InvalidPath,
-      ImportPrefixMissing(_, _) => ErrorKind::ImportPrefixMissing,
-    }
+    ErrorKind::UrlError
```
Comment (ry) [inline cli/deno_error.rs]: I think it would be better to match the name of the JS Error class - `ErrorKind::URIError`.

File: cli/deno_error.rs [hunk 5]
```diff
@@ -252,6 +247,8 @@ impl GetErrorKind for DlopenError {
   }
 }
 
+// NOTE(bartlomieju): seems this is necessary - can't use ErrBox here
+// TODO(bartlomieju): ultimately this should be rewritten to `RuntimeError`?
```
Comment (ry) [inline cli/deno_error.rs]: > TODO: ultimately this should be rewritten to `RuntimeError`?

I think this TODO is in the wrong place. This is not a RuntimeError.

File: cli/deno_error.rs [hunk 6]
```diff
@@ -252,6 +247,8 @@ impl GetErrorKind for DlopenError {
...
 impl GetErrorKind for dyn AnyError {
   fn kind(&self) -> ErrorKind {
```
Comment (ry) [inline cli/deno_error.rs]: I don't think this impl is necessary. I suggest:
```rust
fn kind(&self) -> ErrorKind {
  unreachable!()
}
```

File: cli/js/errors.ts [hunk 7]
```diff
@@ -1,5 +1,18 @@
 // Copyright 2018-2020 the Deno authors. All rights reserved. MIT license.
 
+export function constructError(kind: ErrorKind, msg: string): never {
+  switch (kind) {
+    case ErrorKind.TypeError:
+      throw new TypeError(msg);
+    case ErrorKind.Other:
+      throw new Error(msg);
+    case ErrorKind.UrlError:
+      throw new URIError(msg);
```
Comment (ry) [inline cli/js/errors.ts]: ErrorKind.URIError

File: cli/deno_error.rs (hunk 5 above)
Comment (bartlomieju) [reply cli/deno_error.rs]: That's right, but the downcasts are exactly for runtime errors... Anyway, still iterating on that

File: cli/js/errors.ts (hunk 7 above)
Comment (bartlomieju) [reply cli/js/errors.ts]: done

File: cli/deno_error.rs (hunk 6 above)
Comment (bartlomieju) [reply cli/deno_error.rs]: Not really, tried that and doesn't work. Need to figure this out

File: cli/deno_error.rs (hunk 1 above)
Comment (ry) [inline cli/deno_error.rs]: I would rewrite the TODO items as thus:

```
TODO:
- rename DenoError to OpError?
- rename JSError to RuntimeException. merge V8Exception?
- remove ErrorKind::WouldBlock
- rename ErrorKind::Other. This corresponds to a generic exception thrown as the global `Error` in JS. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error
```

File: cli/deno_error.rs [hunk 8]
```diff
@@ -16,6 +44,38 @@ use std::fmt;
...
+  Interrupted = 15,
+  WriteZero = 16,
+  Other = 17,
+  UnexpectedEof = 18,
+  BadResource = 19,
+  Http = 21,
+  UnixError = 24,
+
+  UrlError = 100,
+  TypeError = 101,
```
Comment (bartlomieju) [inline cli/deno_error.rs]: Note to self; obviously to be changed to be ordered with the rest

**Comment by bartlomieju** [comment]
@ry I'd prefer to land this PR as is if CI goes green. We should be able to rewrite the rest tomorrow.

File: cli/compilers/ts.rs [hunk 9]
```diff
@@ -710,6 +710,7 @@ mod tests {
       .compile_async(mock_state.clone(), &out, TargetLib::Main)
       .await;
     assert!(result.is_ok());
+    eprintln!("asdf {}", result.as_ref().unwrap().code);
```
Comment (ry) [inline cli/compilers/ts.rs]: Remove

Comment (bartlomieju) [reply cli/compilers/ts.rs]: Yep, sry

File: cli/deno_error.rs [hunk 10]
```diff
@@ -414,26 +368,12 @@ mod tests {
...
-    assert_eq!(strip_ansi_codes(&err.to_string()), "error TS2322: Example 1\n\n► deno/tests/complex_diagnostics.ts:19:3\n\n19   values: o => [\n     ~~~~~~\n\nerror TS2000: Example 2\n\n► /foo/bar.ts:129:3\n\n129   values: undefined,\n      ~~~~~~\n\n\nFound 2 errors.\n");
-  }
-
-  #[test]
-  fn test_js_error() {
-    let err = ErrBox::from(js_error());
-    assert_eq!(err.kind(), ErrorKind::JSError);
-    assert_eq!(strip_ansi_codes(&err.to_string()), "error: Error: foo bar\n    at foo (foo_bar.ts:5:17)\n    at qat (bar_baz.ts:6:21)\n    at deno_main.js:2:2");
-  }
-
```
Comment (ry) [inline cli/deno_error.rs]: Why are these tests removed?

**Pull request merged**